From 9fdc27e201c5514241aa4d573523127e70b650b6 Mon Sep 17 00:00:00 2001
From: Techcable <Techcable@techcable.net>
Date: Mon, 24 Apr 2017 11:12:40 -0700
Subject: [PATCH] Better version checking

Checks for outdated versions on startup, and warns on custom versions.

diff --git a/src/main/java/org/bukkit/craftbukkit/Main.java b/src/main/java/org/bukkit/craftbukkit/Main.java
index ad3dd94fa..74ae873a2 100644
--- a/src/main/java/org/bukkit/craftbukkit/Main.java
+++ b/src/main/java/org/bukkit/craftbukkit/Main.java
@@ -15,6 +15,12 @@ import joptsimple.OptionParser;
 import joptsimple.OptionSet;
 import net.minecraft.server.MinecraftServer;
 import org.fusesource.jansi.AnsiConsole;
+// TacoSpigot start
+import net.techcable.tacospigot.TacoVersionInfo;
+import net.techcable.tacospigot.UnknownVersionException;
+import net.techcable.tacospigot.VersionDifference;
+import net.techcable.tacospigot.VersioningException;
+// TacoSpigot end
 
 public class Main {
     public static boolean useJline = true;
@@ -201,6 +207,8 @@ public class Main {
                     useConsole = false;
                 }
 
+                // TacoSpigot start - Actually Check if the version is outdated
+                /*
                 if (Main.class.getPackage().getImplementationVendor() != null && System.getProperty("IReallyKnowWhatIAmDoingISwear") == null) {
                     Date buildDate = new SimpleDateFormat("yyyyMMdd-HHmm").parse(Main.class.getPackage().getImplementationVendor());
 
@@ -213,6 +221,42 @@ public class Main {
                         Thread.sleep(TimeUnit.SECONDS.toMillis(15));
                     }
                 }
+                */
+                // NOTE: Can't use TacoVersioning.currentVersion() since Bukkit.getServer() is null
+                TacoVersionInfo currentVersion = TacoVersionInfo.parse(CraftServer.class.getPackage().getImplementationVersion());
+                VersionDifference versionDifference = null;
+                boolean isCustom = currentVersion.isCustom();
+                try {
+                    versionDifference = currentVersion.compareToLatest();
+                } catch (UnknownVersionException e) {
+                    isCustom = true;
+                } catch (VersioningException | IOException e) {
+                    System.err.println("WARNING: Unable to fetch latest version info!");
+                    e.printStackTrace();
+                }
+                if (isCustom) {
+                    System.err.println("*** Warning, this build (" + currentVersion + ") is custom ***");
+                    System.err.println("*** No support will be offered by the TacoSpigot or Paper authors! ***");
+                    System.err.println("*** Unless this is intended, please download a new build from https://ci.techcable.net/ ***");
+                    System.err.println("*** Server will start in 5 seconds ***");
+                Thread.sleep(TimeUnit.SECONDS.toMillis(5));
+                } else if (versionDifference != null) {
+                    if (versionDifference.isOutdated()) {
+                        System.err.println("*** Error, this build (" + currentVersion + ") is outdated ***");
+                        System.err.println("*** It is " + versionDifference + " the latest version! ***");
+                        System.err.println("*** No support will be offered by the TacoSpigot or Paper authors! ***");
+                        System.err.println("*** Please download a new build as per instructions from https://ci.techcable.net/ ***"); // Paper
+                        System.err.println("*** Server will start in 15 seconds ***");
+                        Thread.sleep(TimeUnit.SECONDS.toMillis(15));
+                    } else if (!versionDifference.isIdentical()) {
+                        System.err.println("*** Warning, this build (" + currentVersion + ") isn't the latest ***");
+                        System.err.println("*** It is " + versionDifference + " the latest version! ***");
+                        System.err.println("*** Unless this is intended, please download a new build from https://ci.techcable.net/ ***");
+                        System.err.println("*** Server will start in 5 seconds ***");
+                        Thread.sleep(TimeUnit.SECONDS.toMillis(5));
+                    }
+                }
+                // TacoSpigot end
 
                 // Spigot Start
                 int maxPermGen = 0; // In kb
-- 
2.13.0

